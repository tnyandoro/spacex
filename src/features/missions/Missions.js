import 'bootstrap/dist/css/bootstrap.min.css';
import React, { useEffect } from 'react';
import {
  Badge,
  Button,
  Container,
  Table,
} from 'react-bootstrap';
import { useDispatch, useSelector } from 'react-redux';
import { exitMission, joinMission, loadMissions } from './missionsReducer';
import './misssion.css';

const Missions = () => {
  const missions = useSelector(({ missionsReducer }) => missionsReducer);
  const dispatch = useDispatch();

  useEffect(() => {
    if (!missions.length) {
      dispatch(loadMissions());
    }
  }, []);
  return (
    <div>
      <Container className="bg-container p-0">
        <Table striped bordered hover>
          <thead>
            <tr>
              <th>Mission</th>
              <th>Description</th>
              <th>Status</th>
              <th>{'  '}</th>
            </tr>
          </thead>
          <tbody>
            {missions.map(({
              id,
              name,
              description,
              reserved,
            }) => (
              <tr key={id}>
                <td><p className="fw-bold">{name}</p></td>
                <td><p>{description}</p></td>
                <td className="align-middle">
                  {reserved && <Badge bg="info">ACTIVE MEMBER</Badge>}
                  {!reserved && <Badge bg="secondary">NOT A MEMBER</Badge>}
                </td>
                <td className="col-2 align-middle text-center">
                  {reserved
                  && (
                  <Button
                    size="sm"
                    variant="danger"
                    onClick={() => dispatch(exitMission(id))}
                  >
                    Leave Mission
                  </Button>
                  )}
                  {!reserved
                  && (
                  <Button
                    size="sm"
                    variant="outline-success"
                    onClick={() => dispatch(joinMission(id))}
                  >
                    Join Mission
                  </Button>
                  )}
                </td>
              </tr>
            ))}
          </tbody>
        </Table>
      </Container>
    </div>
  );
};

export default Missions;
