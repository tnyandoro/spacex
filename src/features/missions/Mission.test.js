import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import Missions from './Missions';
import reducer from './missionsReducer';

describe('Missions Component', () => {
  const initialState = {
    rocketsReducer: [],
    missionsReducer: [],
  };
  const mockStore = configureStore();
  const store = mockStore(initialState);

  it('Missions snapshot test', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Missions />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('mission reducer', () => {
  const payload = { id: 123, name: 'Mock Name', description: 'Mock description' };
  const setMission = payload => ({
    type: 'missionsStore/missions/SET_MISSIONS',
    payload,
  });

  it('Shoud return initial state', () => {
    expect(reducer([], [])).toEqual([]);
  });

  it('Shoud set the first state state', () => {
    expect(reducer([], setMission(payload))).toEqual([payload]);
  });
});
