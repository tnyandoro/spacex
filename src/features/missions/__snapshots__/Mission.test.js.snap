// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions Component Missions snapshot test 1`] = `
<div>
  <div
    className="bg-container p-0 container"
  >
    <table
      className="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            Mission
          </th>
          <th>
            Description
          </th>
          <th>
            Status
          </th>
          <th>
              
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;
